src/storage/src/render/upsert.rs

```diff
@@ -363,27 +378,40 @@ fn stage_input<T, FromTime>(
     }
 }
 
+/// The style of drain we are performing on the stash. `AtTime`-drains cannot
+/// assume that all values have been seen, and must leave tombstones behind for deleted values.
+#[derive(Debug)]
+enum DrainStyle<'a, T> {
+    ToUpper(&'a Antichain<T>),
+    AtTime(T),
+}
+
 /// Helper method for `upsert_inner` used to stage `data` updates
 /// from the input timely edge.
 async fn drain_staged_input<S, G, T, FromTime, E>(
     stash: &mut Vec<(T, UpsertKey, Reverse<FromTime>, Option<UpsertValue>)>,
     commands_state: &mut indexmap::IndexMap<UpsertKey, types::UpsertValueAndSize<Option<FromTime>>>,
     output_updates: &mut Vec<(Result<Row, UpsertError>, T, Diff)>,
     multi_get_scratch: &mut Vec<UpsertKey>,
-    upper: &Antichain<T>,
+    drain_style: DrainStyle<'_, T>,
     error_emitter: &mut E,
     state: &mut UpsertState<'_, S, Option<FromTime>>,
 ) where
     S: UpsertStateBackend<Option<FromTime>>,
     G: Scope,
     T: PartialOrder + Ord + Clone + Debug,
-    FromTime: timely::ExchangeData + Ord + Debug,
+    FromTime: timely::ExchangeData + Ord,
     E: UpsertErrorEmitter<G>,
 {
     stash.sort_unstable();
 
     // Find the prefix that we can emit
-    let idx = stash.partition_point(|(ts, _, _, _)| !upper.less_equal(ts));
+    let idx = stash.partition_point(|(ts, _, _, _)| match &drain_style {
+        DrainStyle::ToUpper(upper) => !upper.less_equal(ts),
+        DrainStyle::AtTime(time) => ts.cmp(time) <= Ordering::Equal,
```

Comment:
Why this cryptic syntax?

```suggestion
        DrainStyle::AtTime(time) => ts < time,
```

Reply:
it was copied from ur sink code that matches on the order